snapshot-labs/envelop
# feat: add endpoints to allow subscriptions management by user

Fix #46 and #57

Allow subscriber to manage their own subscriptions 

# Features

- Add a new `subscriptions` column in the table, to hold the subscriber's subscriptions.  Subscriptions is a list of email type (templateID). See schema.sql
- By default, on create, a subscriber will be subscribed to all email type
- When unsubscribing via the link in the email's footer, the unsubscribe page (envelop-ui) will show a form with a all subscriptions as checkbox, inviting user to select only the type of email he want to receive. By unchecking all options, he will be unsubscribed from all (deleted from db)
- Subscriber can also manage his subscription via his user profile. Same form with checkboxes, but on submit, a message signature will be required to verify wallet ownership.

This PR adds 2 new endpoints:

- `/subscriptionsList`: return a static list of all available email type, used by front end to show the form
- `/subscriber`: return an address' subscriptions 

# How to test 

Most of the changes are on the API side, and made for interaction with UI.
UI will be required for testing:

### testing from envelop-ui, unsubscribe page 

- require this branch https://github.com/snapshot-labs/envelop-ui/pull/3 , for UI interaction
- Boot an instance of envelop-ui, with the branch above
- Boot an instance of snapshot.org, with the branch https://github.com/snapshot-labs/snapshot/pull/3754
- Configure snapshot to use your local instance of envelop in the .env
- Subscribe to email notification via the snapshot UI
- You will receive a verification email
- Verify your email
- With the CLI scripts in envelop repo, send yourself a summary email
- In the email, go to footer, and click on the unsubscribe link
- The unsubscribe page will contains a form
- Submit with various options, and observe the changes in the database (column `subscriptions` should update)
- By unchecking all options and submit, the email should be removed from the database


### testing from snapshop ui, user profile email subscription section 

- require https://github.com/snapshot-labs/snapshot/issues/3886
- Click on your account in the page headr
- in the dropdown, click on email notifications
- in the modal, play with the form, and submit
- Sign the message
- observe changes in the database
-
